src/rdkafka_msgset_reader.c

```diff
@@ -692,8 +741,123 @@ rd_kafka_msgset_reader_msg_v2 (rd_kafka_msgset_reader_t *msetr) {
+                                rktp->rktp_partition,
+                                hdr.Offset, ctrl_data.KeySize);
+
+                rd_kafka_buf_read_i16(rkbuf, &ctrl_data.Version);
+
+                if (ctrl_data.Version != 0) {
+                        rd_rkb_dbg(msetr->msetr_rkb, MSG, "MSG",
+                                "%s [%"PRId32"]: "
+                                "Skipping ctrl msg with "
+                                "unsupported version %"PRId16
+                                " at offset %"PRId64,
+                                rktp->rktp_rkt->rkt_topic->str,
+                                rktp->rktp_partition,
+                                ctrl_data.Version, hdr.Offset);
+                        rd_kafka_buf_skip_to(rkbuf, message_end);
+                        return RD_KAFKA_RESP_ERR_NO_ERROR; /* Continue with next msg */
+                }
+
+                if (unlikely(ctrl_data.KeySize != 4))
+                        rd_kafka_buf_parse_fail(rkbuf,
+                                "%s [%"PRId32"]: "
+                                "Ctrl message at offset %"PRId64
+                                " has invalid key size %"PRId64,
+                                rktp->rktp_rkt->rkt_topic->str,
+                                rktp->rktp_partition,
+                                hdr.Offset, ctrl_data.KeySize);
+
+                rd_kafka_buf_read_i16(rkbuf, &ctrl_data.Type);
+
+                /* Client is uninterested in value of commit marker */
+                rd_kafka_buf_skip(rkbuf, (int32_t)(message_end 
+                        - rd_slice_offset(&rkbuf->rkbuf_reader)));
+
+                switch (ctrl_data.Type) {
+                case RD_KAFKA_CTRL_MSG_COMMIT:
+                        /* always ignore. */
+                        break;
+
+                case RD_KAFKA_CTRL_MSG_ABORT:
+                        if (msetr->msetr_rkb->rkb_rk->rk_conf.isolation_level !=
+                                        RD_KAFKA_READ_COMMITTED)
+                                break;
+
+                        if (unlikely(!msetr->msetr_aborted_txns))
+                                goto unexpected_abort_txn;
+
+                        aborted_txn_start_offset = rd_kafka_aborted_txns_next_aborted_txn_offset_for_pid(
+                                msetr->msetr_aborted_txns, msetr->msetr_v2_hdr->PID);
+
+                        if (unlikely(aborted_txn_start_offset == -1))
+                                goto unexpected_abort_txn;
+
+                        if (unlikely(aborted_txn_start_offset >= hdr.Offset)) {
+                                rd_rkb_log(msetr->msetr_rkb, LOG_ERR, "TXN",
+                                        "%s [%"PRId32"]: "
+                                        "Abort txn ctrl msg bad order "
+                                        "at offset %"PRId64". Expected "
+                                        " before or at %"PRId64,
```

Comment:
Needs to say what we do: do we ignore these messages? do we stop?

Reply:
ignore.
ok.

Reply:
the behavior is kind of undefined. I've changed it to say that messages in aborted transactions may be delivered to the application (which seems likely).